### PR TITLE
Handle encoding binary data separately

### LINKED ISSUE #16989: Bug Report: User defined variables lose binary encoding information
### Overview of the Issue

When using user defined variables in a session, information is lost when the value is passed down the MySQL when interpolated. 

In the reproduction below, it can be seen that the collation for the `@pt` variable is returned as `utf8mb4_0900_ai_ci`. What is happening here is that the bind variables are sent down to `vttablet` and there interpolated, but then the literal string is used without any annotation.

So MySQL ends up running `select collation('foo', rand();`, which means the collation information is lost. We need to make sure that the variable is escaped as a proper binary string. 

### Reproduction Steps

```
mysql> set @p2 = 'foo';
Query OK, 0 rows affected (0.01 sec)

mysql> select @p2;
Field   1:  `@p2`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8mb4_0900_ai_ci (255)
Length:     0
Max_length: 3
Decimals:   0
Flags:      


+------+
| @p2  |
+------+
| foo  |
+------+
1 row in set (0.00 sec)

mysql> select collation(@p2), rand();
Field   1:  `collation(@p2)`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8mb4_0900_ai_ci (255)
Length:     256
Max_length: 18
Decimals:   31
Flags:      

Field   2:  `rand()`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       DOUBLE
Collation:  binary (63)
Length:     23
Max_length: 17
Decimals:   31
Flags:      NOT_NULL BINARY NUM 


+--------------------+-------------------+
| collation(@p2)     | rand()            |
+--------------------+-------------------+
| utf8mb4_0900_ai_ci | 0.048512788534642 |
+--------------------+-------------------+
1 row in set (0.00 sec)
```

### Binary Version

```sh
-
```

### Operating System and Environment details

```sh
-
```

### Log Fragments

```sh
-
```

### PR DESCRIPTION
**Pull request by dbussink**
Binary data should be encoded as such when interpolated, for example in bind variables. Otherwise valuable type information that the value is actually binary is lost.

Also means that we should not mark literals in the parser as binary either. They are strings and there's other introducers like `_binary` that end up marking things as proper binary.

## Related Issue(s)

Fixes #16989 

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by vmg — APPROVED** [review]
These fixes all look great :bow:

File: go/vt/vttablet/tabletserver/tx_engine_test.go [hunk 1]
```diff
@@ -625,43 +625,43 @@ func TestCheckReceivedError(t *testing.T) {
...
 	}, {
 		receivedErr: vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT, "invalid argument"),
 		retryable:   false,
-		expQuery:    `update _vt.redo_state set state = 0, message = 'invalid argument' where dtid = 'aa'`,
+		expQuery:    `update _vt.redo_state set state = 0, message = 'invalid argument' where dtid = _binary'aa'`,
 	}, {
 		receivedErr: sqlerror.NewSQLError(sqlerror.ERLockDeadlock, sqlerror.SSLockDeadlock, "Deadlock found when trying to get lock; try restarting transaction"),
 		retryable:   false,
-		expQuery:    `update _vt.redo_state set state = 0, message = 'Deadlock found when trying to get lock; try restarting transaction (errno 1213) (sqlstate 40001)' where dtid = 'aa'`,
+		expQuery:    `update _vt.redo_state set state = 0, message = 'Deadlock found when trying to get lock; try restarting transaction (errno 1213) (sqlstate 40001)' where dtid = _binary'aa'`,
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/tx_engine_test.go:636]: Is this the side effect or it is necessary to have `_binary` for all varbinary column types?

**Review by harshit-gangal — APPROVED** [review]
Looks great. Just left 1 comment for understanding.

File: go/vt/vttablet/tabletserver/tx_engine_test.go (hunk 1 above)
Comment (dbussink) [reply go/vt/vttablet/tabletserver/tx_engine_test.go:636]: This query is generated with a bind var which now always uses _binary for a binary value. 

Doing this is required to fix the original bug where variables weren't correctly typed and resulted in broken behavior.

**Pull request merged**